hooks/useTransactionSender.tsx

```diff
@@ -1,25 +1,320 @@
+
+      if (isTransactionSuccessful(transactionInfo)) return 'Transaction Succeeded';
+      if (isTransactionUnconfirmed(transactionInfo)) return 'Unable to Confirm Transaction Status';
+      if (isTransactionFailed(transactionInfo)) return 'Transaction Failed';
+
+      return 'Confirming transaction';
+    }
+
+    return `Confirmed ${notificationId ? successfulSignatureCount[notificationId] ?? 0 : 0} of ${
+      transactionInfos.length
+    } transactions`;
+  };
+
+  const generateNotficationBody = (transactionInfos: Array<TransactionInfo<T>>) => {
+    return (
+      <>
+        {transactionInfos.map((transactionInfo, idx) => {
+          const {
+            transaction: { canonicalDescriptor },
+            signature,
+          } = transactionInfo;
+
+          return (
+            <Group align="flex-start" key={idx}>
+              {renderSignatureIcon(transactionInfo)}
+              <Text>
+                <a
+                  href={generateExplorerLink(signature, 'transaction')}
+                  target="_blank"
+                  rel="noreferrer"
+                  onClick={(e) => e.stopPropagation()}
+                  style={{ textDecoration: 'underline', fontWeight: '400' }}
+                >
+                  View Transaction
+                </a>
+                {': '}
+                {canonicalDescriptor ? canonicalDescriptor : `Transaction ${idx + 1}`}
+              </Text>
+            </Group>
+          );
+        })}
+      </>
+    );
+  };
+
+  const renderSignatureIcon = (transactionInfo: TransactionInfo<T>) => {
+    if (isTransactionSuccessful(transactionInfo)) {
+      return (
+        <IconCircleCheck
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-green-filled)"
+        />
+      );
+    } else if (isTransactionFailed(transactionInfo)) {
+      return (
+        <IconCircleX
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-red-filled)"
+        />
+      );
+    } else if (isTransactionUnconfirmed(transactionInfo)) {
+      return (
+        <IconExclamationCircle
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-yellow-filled)"
+        />
+      );
+    }
+
+    return <Loader size="xs" color="var(--mantine-color-dark-2)" />;
+  };
+
+  /**
+   * asynchronously confirm transaction result for each signature and update state when resolved
+   */
+  const confirmTransaction = async (transactionInfos: Array<TransactionInfo<T>>, id: string) => {
+    transactionInfos.forEach(async (transactionInfo) => {
+      const controller = new AbortController();
+      const signal = controller.signal;
+
+      const { blockhash, lastValidBlockHeight } = await connection.getLatestBlockhash();
+      const confirmTransactionPromise = connection.confirmTransaction({
```

Comment:
We could probably give feedback faster by requesting the `processed` commitment level here since `finalized` is the default.

https://github.com/solana-labs/solana-web3.js/blob/77d935221a4805107b20b60ae7c1148725e4e2d0/packages/library-legacy/src/connection.ts#L6664

Reply:
Another option is to use the [getSignatureStatus](https://solana-labs.github.io/solana-web3.js/classes/Connection.html#getSignatureStatus) function. I first went with `confirmTransaction` since that's what was previously used.

Polling `getSignatureStatus` would still give us success + confirmation level, error, or lack of confirmation. Although, I think we have to call another function to actually see the error if non-null and we wanted to display parsed error information. But, that could also be punted to the new frontend work.

Reply:
I don't think it's expensive to setup a WS to pool a txn for lifecycle. However I do wonder if this fits better into overall state manager given there will be listeners for say balance changes or orders filled.

That being said, I'd say we want the OPPORTUNITY to update state optimistically from the users action and result no holds bar. What a developer or application chooses to do with it, is up to them.

Reply:
That's a fair point. I don't have as much experience working with the WS API's, so I defaulted to the single function call. I also didn't put a ton of thought into it since we are probably going to rework everything in the new site anyways 😅 But, is something like the [useMultiAccountSubscription hook](https://github.com/metaDAOproject/meta-dao-frontend/blob/master/hooks/useMultiAccountSubscription.ts) match what you're talking about wrt to WS and listening for updates?

For flexibility, we could also expose an optional commitment parameter that indicates the caller's preferred commitment for determining success.

Reply:
Ok, I was able to change up the logic using a WS (`onSignature` function) — it might be a little cleaner 🧹 Will push in the next revision and see what you think!